providers/aws/provider/provider.go

```diff
@@ -0,0 +1,210 @@
+
+func (s *Service) ParseCLI(req *plugin.ParseCLIReq) (*plugin.ParseCLIRes, error) {
+	flags := req.Flags
+	if flags == nil {
+		flags = map[string]*llx.Primitive{}
+	}
+
+	inventoryConfig := &inventory.Config{
+		Type: "aws",
+	}
+
+	// discovery flags
+	discoverTargets := []string{}
+	if x, ok := flags["discover"]; ok && len(x.Array) != 0 {
+		for i := range x.Array {
+			entry := string(x.Array[i].Value)
+			discoverTargets = append(discoverTargets, entry)
+		}
+	}
+	inventoryConfig.Discover = &inventory.Discovery{Targets: discoverTargets}
+	asset := inventory.Asset{
+		Connections: []*inventory.Config{inventoryConfig},
+		Options:     parseFlagsToOptions(flags),
+	}
+	return &plugin.ParseCLIRes{Asset: &asset}, nil
+}
+
+func parseFlagsToOptions(m map[string]*llx.Primitive) map[string]string {
+	o := make(map[string]string, 0)
+	for k, v := range m {
+		if k == "profile" || k == "region" || k == "role" || k == "endpoint-url" {
+			if val := string(v.Value); val != "" {
+				o[k] = string(v.Value)
+			}
+		}
+	}
+	return o
+}
+
+func (s *Service) Connect(req *plugin.ConnectReq, callback plugin.ProviderCallback) (*plugin.ConnectRes, error) {
+	if req == nil || req.Asset == nil {
+		return nil, errors.New("no connection data provided")
+	}
+
+	conn, err := s.connect(req, callback)
+	if err != nil {
+		return nil, err
+	}
+
+	// // We only need to run the detection step when we don't have any asset information yet.
+	if req.Asset.Platform == nil {
+		if err := s.detect(req.Asset, conn); err != nil {
+			return nil, err
+		}
+	}
+	inventory, err := s.discover(conn)
+	if err != nil {
+		return nil, err
+	}
+	return &plugin.ConnectRes{
+		Id:        uint32(conn.ID()),
+		Name:      conn.Name(),
+		Asset:     req.Asset,
+		Inventory: inventory,
+	}, nil
+}
+
+func (s *Service) connect(req *plugin.ConnectReq, callback plugin.ProviderCallback) (*connection.AwsConnection, error) {
+	if len(req.Asset.Connections) == 0 {
+		return nil, errors.New("no connection options for asset")
+	}
+	asset := req.Asset
+	conf := asset.Connections[0]
+	var conn *connection.AwsConnection
+	var err error
+
+	switch conf.Type {
+	default:
+		s.lastConnectionID++
+		conn, err = connection.NewAwsConnection(s.lastConnectionID, asset, conf)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	var upstream *upstream.UpstreamClient
+	if req.Upstream != nil {
+		upstream, err = req.Upstream.InitClient()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	asset.Connections[0].Id = conn.ID()
+	s.runtimes[conn.ID()] = &plugin.Runtime{
+		Connection:     conn,
+		Resources:      map[string]plugin.Resource{},
+		Callback:       callback,
+		HasRecording:   req.HasRecording,
+		CreateResource: resources.CreateResource,
+		Upstream:       upstream,
+	}
+
+	return conn, err
+}
+
+func (s *Service) detect(asset *inventory.Asset, conn *connection.AwsConnection) error {
+	asset.Id = conn.Conf.Type + "://" + conn.AccountId()
+	asset.Name = conn.Conf.Host
+	asset.Platform = &inventory.Platform{
```

Comment:
Could we re-use `getPlatformForObject` here? It looks quite similar.

Reply:
oh, ya, probably!